### PR TITLE
Schema-registry integration

### PR DESCRIPTION
**Pull request by dankraw**
This PR introduces support for [Confluent Schema Registry](https://github.com/confluentinc/schema-registry) as the next default Avro schema repository. 

Important changes:
* removed support for JSON schema, JSON topics validation & validationDryRun (including changes in `hermes-console`) - from now on schemas are supported for Avro topics only
* removed Zookeeper schema repository implementation (didn't support all needed features)
* upgraded kafka library from 0.8.2.1 to 0.10.0.1 (dependency upgrade only)
* separated `hermes-schema` module for schema-related content (note: DI-factory classes, Avro message wrappers / SerDe are left in `hermes-common`)

### CODE REVIEW

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/di/CommonBinder.java [hunk 1]
```diff
@@ -34,17 +32,16 @@
 import pl.allegro.tech.hermes.common.metric.HermesMetrics;
 import pl.allegro.tech.hermes.common.metric.counter.CounterStorage;
 import pl.allegro.tech.hermes.common.metric.counter.zookeeper.ZookeeperCounterStorage;
+import pl.allegro.tech.hermes.common.schema.AvroCompiledSchemaRepositoryFactory;
+import pl.allegro.tech.hermes.common.schema.SchemaRepositoryFactory;
+import pl.allegro.tech.hermes.schema.SchemaRepository;
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/di/CommonBinder.java]: \[Checkstyle\] ERROR: Unused import \- pl\.allegro\.tech\.hermes\.schema\.SchemaRepository\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java [hunk 2]
```diff
@@ -1,42 +1,73 @@
...
     }
 
     public UnwrappedMessageContent unwrapAvro(byte[] data,
-                                              Topic topic,
-                                              AvroSchemaSource schemaSource) {
-        return topic.isSchemaVersionAwareSerializationEnabled() ?
-                deserialize(data, topic, schemaSource) : schemaSource.tryHard(topic, schema -> avroMessageContentWrapper.unwrapContent(data, schema));
+                                              Topic topic) {
+        return topic.isSchemaVersionAwareSerializationEnabled()
+                ? deserializeSchemaVersionAwarePayload(data, topic)
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java]: \[Checkstyle\] ERROR: '?' should be on the previous line\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java [hunk 3]
```diff
@@ -1,42 +1,73 @@
...
+                : tryDeserializingUsingAnySchemaVersion(data, topic);
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java]: \[Checkstyle\] ERROR: ':' should be on the previous line\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java [hunk 4]
```diff
@@ -1,42 +1,73 @@
...
     }
 
-    private UnwrappedMessageContent deserialize(byte[] data, Topic topic, AvroSchemaSource schemaSource) {
+    private UnwrappedMessageContent deserializeSchemaVersionAwarePayload(byte[] data, Topic topic) {
         SchemaAwarePayload payload = SchemaAwareSerDe.deserialize(data);
-        return avroMessageContentWrapper.unwrapContent(payload.getPayload(), schemaSource.getAvroSchema(topic, payload.getSchemaVersion()));
+        return avroMessageContentWrapper.unwrapContent(payload.getPayload(), schemaRepository.getAvroSchema(topic, payload.getSchemaVersion()));
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 144\)\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java [hunk 5]
```diff
@@ -1,42 +1,73 @@
...
+            logger.info("Trying to find schema online for message for topic {}", topic.getQualifiedName());
+            return tryDeserializingUsingAnySchemaVersion(data, topic, true);
+        }
+    }
+
+    private UnwrappedMessageContent tryDeserializingUsingAnySchemaVersion(byte[] data, Topic topic, boolean online) {
+        List<SchemaVersion> versions = schemaRepository.getVersions(topic, online);
+        for (SchemaVersion version : versions) {
+            try {
+                return avroMessageContentWrapper.unwrapContent(data, online? schemaRepository.getKnownAvroSchemaVersion(topic, version)
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java]: \[Checkstyle\] ERROR: '?' is not preceded with whitespace\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java [hunk 6]
```diff
@@ -1,42 +1,73 @@
...
+                        : schemaRepository.getAvroSchema(topic, version));
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/MessageContentWrapper.java]: \[Checkstyle\] ERROR: ':' should be on the previous line\.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/schema/SchemaSourceClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    public SchemaSourceClientFactory(ConfigFactory configFactory) {
+        this.configFactory = configFactory;
+    }
+
+    @Override
+    public SchemaSourceClient provide() {
+
+        ClientConfig config = new ClientConfig()
+                .property(ClientProperties.READ_TIMEOUT, configFactory.getIntProperty(Configs.SCHEMA_REPOSITORY_HTTP_READ_TIMEOUT_MS))
+                .property(ClientProperties.CONNECT_TIMEOUT, configFactory.getIntProperty(Configs.SCHEMA_REPOSITORY_HTTP_CONNECT_TIMEOUT_MS));
```
Comment (sputnikci) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/schema/SchemaSourceClientFactory.java]: \[Checkstyle\] INFO: Line is longer than 140 characters \(found 141\)\.

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/schema/SchemaSourceService.java [hunk 8]
```diff
@@ -4,58 +4,63 @@
...
+import pl.allegro.tech.hermes.schema.SchemaSourceClient;
+import pl.allegro.tech.hermes.schema.SchemaVersion;
 
 import java.util.Optional;
 
-import static pl.allegro.tech.hermes.api.ContentType.AVRO;
 import static pl.allegro.tech.hermes.api.TopicName.fromQualifiedName;
 
 @Component
 public class SchemaSourceService {
```
Comment (adamdubiel) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/schema/SchemaSourceService.java]: i wonder - we use term `schema source` everywhere and i don't know if it's a proper one - schema source sounds like source of schema aka schema repository, while what we mean is schema itself (aka raw schema in text format). Maybe it's a good time to change this?

File: hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
+package pl.allegro.tech.hermes.schema.schemarepo;
```
Comment (adamdubiel) [inline hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java]: why is it in java and not groovy? this is weird, as other tests are in groovy in this module

**Comment by adamdubiel** [comment]
I'm mostly confused about the naming, but all in all - good job :)

File: hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java (hunk 9 above)
Comment (dankraw) [inline hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java]: I've been reusing old tests, transforming them to groovy shouldn't take long. :)

File: hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/schema/SchemaSourceService.java (hunk 8 above)
Comment (dankraw) [inline hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/schema/SchemaSourceService.java]: Agreed, I'll rename it to `RawSchema`.

File: hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java (hunk 9 above)
Comment (dankraw) [inline hermes-schema/src/test/java/pl/allegro/tech/hermes/schema/schemarepo/SchemaRepoSchemaSourceClientTest.java]: OK, all `hermes-schema` tests are now groovy :)

**Pull request merged**